### ISSUE TITLE
[Perf] AR CUDA graph mode: remaining bubbles after 10x eager-to-graph speedup

### ISSUE THREAD

**Issue by TaffyOfficial** [vllm-project/vllm-omni#3730]
# [Perf] AR CUDA graph mode: remaining bubbles after 10x eager-to-graph speedup

## Background

Local AR profiling shows that enabling CUDA graph / compile is the dominant framework-level optimization for the HunyuanImage3 AR path. The current graph-mode trace is much faster than eager, but still has two visible residual bubbles:

- first real graph-mode prefill has large MoE cold-start bubbles
- steady decode still has ~4 ms/step host-side gap, with prefix-cache DtoH copies as a likely contributor

Source traces used locally:

- graph mode: `D:\vllm-omni\workflow-starter\.scratch\traces_graph_baseline\trace_rank0.json`
- same-window eager baseline: `D:\vllm-omni\workflow-starter\.scratch\traces_baseline\trace_rank0.json`
- external no-graph L20X TP2 trace: `D:\vllm-omni\ar打点\ar_l20x_tp2\trace_rank0.json`

## Current measurements

Same-window A/B, 26 decode steps:

| Metric | Eager baseline | CUDA graph baseline |
| --- | ---: | ---: |
| Decode wall time | 119.51 ms/step | 11.13 ms/step |
| GPU union busy | 85.13 ms/step | 6.98 ms/step |
| Bubble / host gap | 34.38 ms/step | 4.14 ms/step |
| `cudaLaunchKernel` | 1512.0 calls/step | 26.7 calls/step |
| `cudaGraphLaunch` | 0 calls/step | 1.0 call/step |
| `cudaMemcpyAsync` runtime | 1.55 ms/step | 6.76 ms/step |

External no-graph L20X TP2 trace is consistent: decode is ~113.2 ms/step, with TP communication dominating the apparent GPU time.

## Top GPU groups per decode step

| Operator group | Eager ms/step | Graph ms/step | Delta |
| --- | ---: | ---: | ---: |
| TP all-reduce / comm | 70.84 | 0.86 | -69.98 |
| attention | 4.66 | 1.09 | -3.57 |
| fused MoE | 4.88 | 2.81 | -2.07 |
| elementwise / cat | 1.62 | 0.09 | -1.53 |
| rms norm | 0.39 | 0.29 | -0.10 |
| reduce | 0.32 | 0.17 | -0.16 |
| DtoD memcpy | 0.17 | 0.004 | -0.16 |
| routing align | 0.086 | 0.071 | -0.015 |
| KV cache | 0.061 | 0.048 | -0.013 |
| GEMM / projection | 2.17 | 2.22 | +0.06 |

Interpretation: graph mode is not making every kernel intrinsically faster. The large win comes from removing launch-bound eager execution and TP rank arrival skew. In eager mode, all-reduce CUDA time includes a lot of collective waiting because ranks arrive at all-reduce at different times. Graph replay makes rank execution much more deterministic, so the apparent all-reduce time collapses.

## Proposed work

I want to make the AR CUDA graph path production-ready when prefix caching is enabled. The concrete framework-level work I propose is:

1. Treat CUDA graph / compile as the intended AR fast path.
   - Remove any accidental `enforce_eager=True` from the HunyuanImage3 AR profiling/deploy path.
   - Make the graph capture sizes explicit for the common decode shapes we care about.
   - Keep eager as a debugging fallback, not the performance path.

2. Deduplicate `hidden_states` GPU-to-CPU staging.
   - Create one per-step `hidden_states_cpu` staging tensor in the AR runner.
   - Reuse that same CPU tensor for prefix-cache update, prefix-cache merge, and pooler/downstream payload construction.
   - Add optional `hidden_states_cpu` parameters to the prefix-cache APIs while preserving the current fallback behavior when the caller does not pass a staging tensor.

3. Make the remaining prefix-cache DtoH copy asynchronous.
   - Replace direct `detach().cpu().contiguous()` copies on the hot path with reusable pinned CPU staging buffers.
   - Use non-blocking copy on a copy stream and track readiness with an event.
   - Wait only at the point where CPU code actually needs to read the staged tensor.

4. Move MoE cold-start cost into warmup/capture.
   - Exercise the MoE path during CUDA graph warmup for the relevant capture sizes.
   - Ensure MoE autotune/workspace/lazy initialization is paid before the first real request.

Out of scope for this issue: model architecture changes, quantization, speculative decoding, and algorithm-level changes. Those may help later, but this issue is specifically about the framework overhead still visible after enabling graph mode.

### 1. First-step MoE cold start

Graph-mode prefill: ~757.6 ms wall, but only ~25.6 ms GPU busy. Two large ~300 ms bubbles appear around the first `vllm::moe_forward_shared` / `fused_moe_kernel` path.

Likely cause: first real graph-mode MoE path pays lazy initialization, autotune, workspace allocation, or graph/capture preparation cost.

Suggested framework-level fix:

- add or extend graph warmup so MoE paths are exercised before the first real request
- cover common `cudagraph_capture_sizes` / batch shapes during warmup
- ensure MoE autotune/workspace setup is completed outside the first measured request

### 2. Steady decode host gap with prefix-cache DtoH

Graph-mode steady decode is ~10.8-11.1 ms/step. GPU union busy is ~6.7 ms/step, leaving ~4 ms/step of gap. `cudaGraphLaunch` itself is ~1.0-1.4 ms p50-ish per step, but the trace also shows costly `cudaMemcpyAsync` runtime and small DtoH copies.

Likely source in code:

- `vllm_omni/core/prefix_cache.py::_coerce_to_cpu_tensor()` uses `detach().cpu().contiguous()`
- `vllm_omni/worker/gpu_ar_model_runner.py` also copies `hidden_states` to CPU for pooler/downstream payload construction

If prefix caching must remain enabled, the optimization should be to avoid multiple blocking pageable DtoH copies in the same step.

Suggested fix direction:

1. Create one per-step `hidden_states_cpu` staging tensor in the AR runner.
2. Pass that CPU tensor into prefix-cache update / prefix-cache merge / pooler payload construction.
3. If no staging tensor is passed, keep existing fallback behavior for compatibility.
4. Move the single remaining DtoH copy to pinned memory + non-blocking copy, guarded by a copy event.
5. Optionally batch prefix-cache writes at block boundaries instead of writing every decode token.

## Expected impact

- `enforce_eager=False` / CUDA graph is already measured at ~10x decode speedup over eager.
- MoE warmup should mostly improve first-token / first-request latency by removing the ~600 ms cold-start bubble.
- Prefix-cache DtoH de-duplication and pinned async copy should reduce the remaining graph-mode steady decode gap, especially when prefix caching is required.

## Notes

This issue is focused on framework-level changes only. It intentionally avoids model architecture changes, quantization, speculative decoding, or algorithm-level changes.

**Comment by TaffyOfficial** [comment]
﻿I opened a first PR for the prefix-cache part of this issue: https://github.com/vllm-project/vllm-omni/pull/3734

This PR addresses the duplicate hidden-state CPU staging that showed up after enabling the AR CUDA graph path.

## How the optimization point was found from profiling

The important part is that this was not inferred from code alone. The profiling flow was:

1. Compare eager vs CUDA graph first.

   The eager trace was about 110 ms/step, while the CUDA graph trace was about 10-11 ms/step. That established CUDA graph/compile as the intended AR fast path. After that, the remaining graph-mode issue became the smaller steady-state host bubble, not the old eager launch storm.

2. Inspect the remaining graph-mode decode bubble.

   In graph mode, the per-step timeline still had a recurring host-side gap around graph replay/output handling. One recurring event inside that region was:

   ```text
   Memcpy DtoH (Device -> Pageable)
   ```

   The GPU-side copy duration itself was tiny, but the associated CPU runtime path was expensive enough to be visible in the step bubble. That is the characteristic shape of a small pageable DtoH copy becoming a synchronization point.

3. Attribute the DtoH copies back to the CPU stack.

   The trace grouped those copies under `aten::copy_` / `aten::_to_copy`, and stack attribution pointed at the prefix-cache CPU materialization path:

   ```text
   vllm_omni/core/prefix_cache.py::_coerce_to_cpu_tensor()
       -> hidden_states.detach().cpu().contiguous()
   ```

   That means prefix-cache update/merge was materializing GPU hidden states on CPU during the AR step.

4. Cross-check the runner source for another consumer of the same data.

   After locating the prefix-cache copy, I checked the surrounding AR runner path. The GPU AR runner also constructs a CPU hidden-state payload for pooler/downstream use. That path was independently materializing the same hidden-state slice with a separate CPU copy.

   In other words, for the same step, the same logical data was needed by multiple consumers:

   - prefix-cache update
   - prefix-cache merge
   - downstream/pooler payload construction

   All of them operate on the same per-step `hidden_states[:num_tokens_unpadded]` slice. The pre-PR shape was effectively:

   ```text
   hidden_states on GPU
      -> prefix-cache update does its own .detach().cpu().contiguous()
      -> prefix-cache merge may do its own .detach().cpu().contiguous()
      -> runner downstream payload does its own .detach().to("cpu").contiguous()
   ```

5. Turn the profiling observation into the lowest-risk framework change.

   Since the consumers need the same CPU representation of the same hidden-state slice, the first safe optimization is not pinned memory or async copies yet. The first safe optimization is to centralize copy ownership:

   ```text
   hidden_states on GPU
      -> one runner-owned CPU staging copy per step
      -> pass hidden_states_cpu to prefix-cache update
      -> pass hidden_states_cpu to prefix-cache merge
      -> reuse hidden_states_cpu for downstream payload construction
   ```

   This preserves semantics because it does not change the values, cache layout, prefix-cache policy, or downstream payload. It only avoids repeated materialization of the same GPU tensor slice on CPU.

## What #3734 changes

- Adds optional `hidden_states_cpu` parameters to the prefix-cache update/merge APIs while preserving fallback behavior for existing callers.
- Adds `hidden_states_cpu` to the GPU AR runner execution state.
- Stages one CPU copy in `GPUARModelRunner.execute_model()` when prefix cache is enabled on the last PP rank.
- Reuses that same CPU tensor for prefix-cache update, prefix-cache merge, and downstream payload construction.
- Validates pre-staged tensors in `OmniTensorPrefixCache` and raises `RuntimeError` for malformed inputs instead of silently recopying.

The validation is intentional: if a caller passes a staged tensor that is not CPU, not contiguous, has the wrong hidden size, or is too short for the requested slice, that is a programming error. Silently falling back to another `.cpu()` would hide the performance bug again.

This PR intentionally does not add pinned memory, copy streams, events, block-boundary batching, or GPU-resident cache yet. It only centralizes copy ownership so those follow-up optimizations have a single place to hook into.

## Profiling validation

Workload:

- HunyuanImage3 full `img2img` AR-to-DiT e2e
- `enable_prefix_caching=true`
- AR CUDA graph enabled
- AR stage 0 torch profiler
- DiT reduced to 1 step to isolate AR prefix-cache overhead
- Same prompt, image, seed, and temporary deploy config for base vs PR

Result:

| Metric | Base | PR | Delta |
| --- | ---: | ---: | ---: |
| rank0 `Memcpy DtoH (Device -> Pageable)` | 288 calls / 793.7 us | 96 calls / 258.9 us | -192 calls, about -67% |
| rank1 `Memcpy DtoH (Device -> Pageable)` | 288 calls / 783.2 us | 96 calls / 247.6 us | -192 calls, about -68% |
| rank0 `aten::copy_` count | 3297 | 3201 | -96 |
| rank1 `aten::copy_` count | 3297 | 3201 | -96 |
| profiled e2e wall time | 5.456 s | 5.355 s | -0.101 s, about -1.9% |

The most important evidence is the DtoH count drop: the PR removes duplicate hidden-state pageable DtoH staging in the real AR-to-DiT path. The e2e wall-time delta is intentionally modest because this PR does not try to fix graph launch overhead, AR compute/comm, DiT time, profiler overhead, or MoE cold start.

## Accuracy validation

Using the HunyuanImage3 accuracy metrics/reference with prefix cache enabled:

| Config | COT sim | Prefix match | CLIP | SSIM | PSNR | Result |
| --- | ---: | ---: | ---: | ---: | ---: | --- |
| `enforce_eager=true` | 0.9644 | 29 | 93.97 | 0.4250 | 14.32 | Pass |
| `enforce_eager=false` | 0.9951 | 70 | 92.33 | 0.4213 | 14.63 | Pass |

Official thresholds used by the test: COT sim >= 0.9, prefix match >= 10, CLIP >= 85, SSIM >= 0.20, PSNR >= 11.0.

## Unit/static checks

```text
ruff check vllm_omni/core/prefix_cache.py vllm_omni/worker/gpu_ar_model_runner.py tests/core/test_prefix_cache.py
All checks passed!

pytest tests/core/test_prefix_cache.py -q
19 passed, 18 warnings in 1.05s
```

## Remaining follow-ups

This PR handles the first low-risk step: remove duplicate hidden-state CPU staging. The remaining items from this issue are still open:

1. Move the single remaining DtoH copy to pinned memory + non-blocking copy once ownership is centralized.
2. Consider batching prefix-cache writes where possible.
3. Move the first-request MoE cold-start cost into warmup/capture.

**Comment by TaffyOfficial** [comment]
@gcanlin
